### PR TITLE
[CI:DOCS]libpod endpoints

### PR DESCRIPTION
**Pull request by baude**
small refactor for libpod inclusion and began adding endpoints.

Signed-off-by: baude <bbaude@redhat.com>

### CODE REVIEW

File: pkg/api/handlers/generic/containers.go [hunk 1]
```diff
@@ -0,0 +1,306 @@
+package generic
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+	"github.com/containers/libpod/pkg/api/handlers/utils"
```
Comment (rhatdan) [inline pkg/api/handlers/generic/containers.go]: Wrong place

File: pkg/api/handlers/generic/containers.go [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+		Force bool `schema:"force"`
+		Vols  bool `schema:"v"`
+		Link  bool `schema:"link"`
+	}{
+		// override any golang type defaults
+	}
+
+	if err := decoder.Decode(&query, r.URL.Query()); err != nil {
+		handlers.Error(w, http.StatusText(http.StatusBadRequest), http.StatusBadRequest,
+			errors.Wrapf(err, "Failed to parse parameters for %s", r.URL.String()))
```
Comment (rhatdan) [inline pkg/api/handlers/generic/containers.go]: No capatilized errors.

File: pkg/api/handlers/generic/containers.go [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+		return
+	}
+	if query.Link {
+		handlers.Error(w, http.StatusText(http.StatusBadRequest), http.StatusBadRequest,
+			errors.New("link is not supported"))
```
Comment (rhatdan) [inline pkg/api/handlers/generic/containers.go]: Can this be a constant.

File: pkg/api/handlers/generic/containers.go [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+
+	containers, err := runtime.GetAllContainers()
+	if err != nil {
+		handlers.InternalServerError(w, err)
+		return
+	}
+
+	infoData, err := runtime.Info()
+	if err != nil {
+		handlers.InternalServerError(w, errors.Wrapf(err, "Failed to obtain system info"))
```
Comment (rhatdan) [inline pkg/api/handlers/generic/containers.go]: No capitalized errors.

File: pkg/api/handlers/generic/containers.go [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+	con, err := utils.KillContainer(w, r)
+	if err != nil {
+		return
+	}
+	// the kill behavior for docker differs from podman in that they appear to wait
+	// for the Container to croak so the exit code is accurate immediately after the
+	// kill is sent.  libpod does not.  but we can add a wait here only for the docker
+	// side of things and mimic that behavior
+	if _, err = con.Wait(); err != nil {
+		handlers.Error(w, "Something went wrong.", http.StatusInternalServerError, errors.Wrapf(err, "failed to wait for Container %s", con.ID()))
```
Comment (rhatdan) [inline pkg/api/handlers/generic/containers.go]: No capitalized errors.

File: pkg/api/handlers/generic/containers.go [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+		Tail:       tail,
+		Timestamps: query.Timestamps,
+	}
+
+	var wg sync.WaitGroup
+	options.WaitGroup = &wg
+
+	logChannel := make(chan *logs.LogLine, tail+1)
+	if err := runtime.Log([]*libpod.Container{ctnr}, options, logChannel); err != nil {
+		handlers.InternalServerError(w, errors.Wrapf(err, "Failed to obtain logs for Container '%s'", name))
```
Comment (rhatdan) [inline pkg/api/handlers/generic/containers.go]: No capitalized errors.

File: pkg/api/handlers/generic/containers.go [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+				header[0] = 1
+			case "stderr":
+				if !query.Stderr {
+					continue
+				}
+				header[0] = 2
+			default:
+				// Logging and moving on is the best we can do here. We may have already sent
+				// a Status and Content-Type to client therefore we can no longer report an error.
+				log.Infof("Unknown Device type '%s' in log file from Container %s", line.Device, ctnr.ID())
```
Comment (rhatdan) [inline pkg/api/handlers/generic/containers.go]: Nit no caps.

File: pkg/api/handlers/generic/containers_create.go [hunk 8]
```diff
@@ -1,8 +1,9 @@
-package handlers
+package generic
 
 import (
 	"encoding/json"
 	"fmt"
+	"github.com/containers/libpod/pkg/api/handlers"
```
Comment (rhatdan) [inline pkg/api/handlers/generic/containers_create.go]: Needs line split

File: pkg/api/handlers/generic/containers_create.go [hunk 9]
```diff
@@ -23,30 +24,30 @@ import (
...
 	}
 	if err := decoder.Decode(&query, r.URL.Query()); err != nil {
-		Error(w, http.StatusText(http.StatusBadRequest), http.StatusBadRequest,
+		handlers.Error(w, http.StatusText(http.StatusBadRequest), http.StatusBadRequest,
 			errors.Wrapf(err, "Failed to parse parameters for %s", r.URL.String()))
 		return
 	}
 	if err := json.NewDecoder(r.Body).Decode(&input); err != nil {
-		Error(w, "Something went wrong.", http.StatusInternalServerError, errors.Wrap(err, "Decode()"))
+		handlers.Error(w, "Something went wrong.", http.StatusInternalServerError, errors.Wrap(err, "Decode()"))
```
Comment (rhatdan) [inline pkg/api/handlers/generic/containers_create.go]: No capitalized errors.

File: pkg/api/handlers/generic/containers_create.go [hunk 10]
```diff
@@ -23,30 +24,30 @@ import (
...
 		return
 	}
 
 	newImage, err := runtime.ImageRuntime().NewFromLocal(input.Image)
 	if err != nil {
-		Error(w, "Something went wrong.", http.StatusInternalServerError, errors.Wrap(err, "NewFromLocal()"))
+		handlers.Error(w, "Something went wrong.", http.StatusInternalServerError, errors.Wrap(err, "NewFromLocal()"))
```
Comment (rhatdan) [inline pkg/api/handlers/generic/containers_create.go]: No capitalized errors.

File: pkg/api/handlers/generic/containers_create.go [hunk 11]
```diff
@@ -23,30 +24,30 @@ import (
...
 		return
 	}
 	cc, err := makeCreateConfig(input, newImage)
 	if err != nil {
-		Error(w, "Something went wrong.", http.StatusInternalServerError, errors.Wrap(err, "makeCreatConfig()"))
+		handlers.Error(w, "Something went wrong.", http.StatusInternalServerError, errors.Wrap(err, "makeCreatConfig()"))
```
Comment (rhatdan) [inline pkg/api/handlers/generic/containers_create.go]: No capitalized errors.

File: pkg/api/handlers/generic/containers_create.go [hunk 12]
```diff
@@ -65,7 +66,7 @@ func CreateContainer(w http.ResponseWriter, r *http.Request) {
 			//s.WriteResponse(w, http.StatusInternalServerError, fmt.Sprintf("logger: no log driver named '%s' is registered", input.HostConfig.LogConfig.Type))
 			return
 		}
-		Error(w, "Something went wrong.", http.StatusInternalServerError, errors.Wrap(err, "CreateContainerFromCreateConfig()"))
+		handlers.Error(w, "Something went wrong.", http.StatusInternalServerError, errors.Wrap(err, "CreateContainerFromCreateConfig()"))
```
Comment (rhatdan) [inline pkg/api/handlers/generic/containers_create.go]: No capitalized errors.

File: pkg/api/handlers/generic/containers_stats.go [hunk 13]
```diff
@@ -27,24 +28,24 @@ func StatsContainer(w http.ResponseWriter, r *http.Request) {
 		Stream: true,
 	}
 	if err := decoder.Decode(&query, r.URL.Query()); err != nil {
-		Error(w, "Something went wrong.", http.StatusBadRequest, errors.Wrapf(err, "Failed to parse parameters for %s", r.URL.String()))
+		handlers.Error(w, "Something went wrong.", http.StatusBadRequest, errors.Wrapf(err, "Failed to parse parameters for %s", r.URL.String()))
```
Comment (rhatdan) [inline pkg/api/handlers/generic/containers_stats.go]: No capitalized errors.

File: pkg/api/handlers/generic/containers_stats.go [hunk 14]
```diff
@@ -56,7 +57,7 @@ func StatsContainer(w http.ResponseWriter, r *http.Request) {
 
 	stats, err := ctnr.GetContainerStats(&libpod.ContainerStats{})
 	if err != nil {
-		InternalServerError(w, errors.Wrapf(err, "Failed to obtain Container %s stats", name))
+		handlers.InternalServerError(w, errors.Wrapf(err, "Failed to obtain Container %s stats", name))
```
Comment (rhatdan) [inline pkg/api/handlers/generic/containers_stats.go]: No capitalized errors.

File: pkg/api/handlers/generic/containers_top.go [hunk 15]
```diff
@@ -21,40 +22,40 @@ func TopContainer(w http.ResponseWriter, r *http.Request) {
...
 
 	state, err := ctnr.State()
 	if err != nil {
-		InternalServerError(w, err)
+		handlers.InternalServerError(w, err)
 		return
 	}
 	if state != define.ContainerStateRunning {
-		ContainerNotRunning(w, name, errors.Errorf("Container %s must be running to perform top operation", name))
+		handlers.ContainerNotRunning(w, name, errors.Errorf("Container %s must be running to perform top operation", name))
```
Comment (rhatdan) [inline pkg/api/handlers/generic/containers_top.go]: No capitalized errors.

File: pkg/api/handlers/generic/images.go [hunk 16]
```diff
@@ -0,0 +1,396 @@
+package generic
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/containers/libpod/pkg/api/handlers"
```
Comment (rhatdan) [inline pkg/api/handlers/generic/images.go]: Wrong place

File: pkg/api/handlers/generic/images.go [hunk 17]
```diff
@@ -0,0 +1,396 @@
...
+
+	var (
+		err    error
+		writer io.Writer
+	)
+	quiet := false
+	if len(r.Form.Get("quiet")) > 0 {
+		quiet, err = strconv.ParseBool(r.Form.Get("quiet"))
+		if err != nil {
+			handlers.Error(w, "Something went wrong.", http.StatusInternalServerError, errors.Wrap(err, "bad value for quiet"))
```
Comment (rhatdan) [inline pkg/api/handlers/generic/images.go]: No capitalized errors.

File: pkg/api/handlers/generic/images.go [hunk 18]
```diff
@@ -0,0 +1,396 @@
...
+			return
+		}
+	}
+	f, err := ioutil.TempFile("", "api_load.tar")
+	if err != nil {
+		handlers.Error(w, "Something went wrong.", http.StatusInternalServerError, errors.Wrap(err, "failed to create tempfile"))
```
Comment (rhatdan) [inline pkg/api/handlers/generic/images.go]: No capitalized errors.

File: pkg/api/handlers/generic/images.go [hunk 19]
```diff
@@ -0,0 +1,396 @@
...
+		return
+	}
+	if err := saveFromBody(f, r); err != nil {
+		handlers.Error(w, "Something went wrong.", http.StatusInternalServerError, errors.Wrap(err, "failed to write temporary file"))
```
Comment (rhatdan) [inline pkg/api/handlers/generic/images.go]: No capitalized errors.

File: pkg/api/handlers/generic/images.go [hunk 20]
```diff
@@ -0,0 +1,396 @@
...
+		return
+	}
+	id, err := runtime.LoadImage(r.Context(), "", f.Name(), writer, "")
+	if err != nil {
+		handlers.Error(w, "Something went wrong.", http.StatusInternalServerError, errors.Wrap(err, "failed to load image"))
```
Comment (rhatdan) [inline pkg/api/handlers/generic/images.go]: No capitalized errors.

File: pkg/api/handlers/generic/images.go [hunk 21]
```diff
@@ -0,0 +1,396 @@
...
+	})
+}
+
+func ExportImage(w http.ResponseWriter, r *http.Request) {
+	runtime := r.Context().Value("runtime").(*libpod.Runtime)
+
+	name := mux.Vars(r)["name"]
+	newImage, err := runtime.ImageRuntime().NewFromLocal(name)
+	if err != nil {
+		handlers.ImageNotFound(w, name, errors.Wrapf(err, "Failed to find image %s", name))
```
Comment (rhatdan) [inline pkg/api/handlers/generic/images.go]: No capitalized errors.

File: pkg/api/handlers/generic/images.go [hunk 22]
```diff
@@ -0,0 +1,396 @@
...
+		return
+	}
+	tmpfile, err := ioutil.TempFile("", "api.tar")
+	if err != nil {
+		handlers.Error(w, "Something went wrong.", http.StatusInternalServerError, errors.Wrap(err, "unable to create tempfile"))
```
Comment (rhatdan) [inline pkg/api/handlers/generic/images.go]: No capitalized errors.

File: pkg/api/handlers/generic/images.go [hunk 23]
```diff
@@ -0,0 +1,396 @@
...
+		return
+	}
+	if err := tmpfile.Close(); err != nil {
+		handlers.Error(w, "Something went wrong.", http.StatusInternalServerError, errors.Wrap(err, "unable to close tempfile"))
```
Comment (rhatdan) [inline pkg/api/handlers/generic/images.go]: No capitalized errors.

File: pkg/api/handlers/generic/images.go [hunk 24]
```diff
@@ -0,0 +1,396 @@
...
+		return
+	}
+	if err := newImage.Save(r.Context(), name, "docker-archive", tmpfile.Name(), []string{}, false, false); err != nil {
+		handlers.Error(w, "Something went wrong.", http.StatusInternalServerError, errors.Wrap(err, "failed to save image"))
```
Comment (rhatdan) [inline pkg/api/handlers/generic/images.go]: No capitalized errors.

File: pkg/api/handlers/generic/images.go [hunk 25]
```diff
@@ -0,0 +1,396 @@
...
+		return
+	}
+	rdr, err := os.Open(tmpfile.Name())
+	if err != nil {
+		handlers.Error(w, "Something went wrong.", http.StatusInternalServerError, errors.Wrap(err, "failed to read the exported tarfile"))
```
Comment (rhatdan) [inline pkg/api/handlers/generic/images.go]: No capitalized errors.

File: pkg/api/handlers/libpod/containers.go [hunk 26]
```diff
@@ -0,0 +1,125 @@
+package libpod
+
+import (
+	"github.com/containers/libpod/cmd/podman/shared"
```
Comment (rhatdan) [inline pkg/api/handlers/libpod/containers.go]: Wrong place.

File: pkg/api/handlers/libpod/images.go [hunk 27]
```diff
@@ -0,0 +1,91 @@
+package libpod
+
+import (
+	"github.com/containers/libpod/libpod"
+	"github.com/containers/libpod/pkg/api/handlers"
+	"github.com/gorilla/mux"
+	"github.com/pkg/errors"
+	"net/http"
```
Comment (rhatdan) [inline pkg/api/handlers/libpod/images.go]: Wrong place

File: pkg/api/handlers/libpod/pods.go [hunk 28]
```diff
@@ -1,8 +1,9 @@
-package handlers
+package libpod
 
 import (
 	"encoding/json"
 	"fmt"
+	"github.com/containers/libpod/pkg/api/handlers"
```
Comment (rhatdan) [inline pkg/api/handlers/libpod/pods.go]: Wrong place

File: pkg/api/handlers/utils/containers.go [hunk 29]
```diff
@@ -0,0 +1,104 @@
+package utils
+
+import (
+	"fmt"
+	"github.com/containers/libpod/pkg/api/handlers"
```
Comment (rhatdan) [inline pkg/api/handlers/utils/containers.go]: Wrong place

File: pkg/api/handlers/generic/containers.go (hunk 2 above)
Comment (baude) [reply pkg/api/handlers/generic/containers.go]: unfortunately they are in the API we are being consistent with.

File: pkg/api/handlers/generic/containers.go (hunk 4 above)
Comment (baude) [reply pkg/api/handlers/generic/containers.go]: to be consistent with the other api, we must use caps.

File: pkg/api/handlers/generic/images.go (hunk 25 above)
Comment (baude) [reply pkg/api/handlers/generic/images.go]: to be consistent with the other api, we must use caps.

File: pkg/api/handlers/generic/containers.go (hunk 6 above)
Comment (baude) [reply pkg/api/handlers/generic/containers.go]: to be consistent with the other api, we must use caps.

File: pkg/api/handlers/generic/containers_stats.go (hunk 13 above)
Comment (baude) [reply pkg/api/handlers/generic/containers_stats.go]: to be consistent with the other api, we must use caps.

File: pkg/api/handlers/generic/containers_top.go (hunk 15 above)
Comment (baude) [reply pkg/api/handlers/generic/containers_top.go]: to be consistent with the other api, we must use caps.

File: pkg/api/handlers/generic/containers_stats.go (hunk 14 above)
Comment (baude) [reply pkg/api/handlers/generic/containers_stats.go]: to be consistent with the other api, we must use caps.

File: pkg/api/handlers/generic/images.go (hunk 17 above)
Comment (baude) [reply pkg/api/handlers/generic/images.go]: to be consistent with the other api, we must use caps.

File: pkg/api/handlers/generic/images.go (hunk 18 above)
Comment (baude) [reply pkg/api/handlers/generic/images.go]: to be consistent with the other api, we must use caps.

File: pkg/api/handlers/generic/images.go (hunk 20 above)
Comment (baude) [reply pkg/api/handlers/generic/images.go]: to be consistent with the other api, we must use caps.

File: pkg/api/handlers/generic/images.go (hunk 19 above)
Comment (baude) [reply pkg/api/handlers/generic/images.go]: to be consistent with the other api, we must use caps.

File: pkg/api/handlers/generic/images.go (hunk 21 above)
Comment (baude) [reply pkg/api/handlers/generic/images.go]: to be consistent with the other api, we must use caps.

File: pkg/api/handlers/generic/images.go (hunk 23 above)
Comment (baude) [reply pkg/api/handlers/generic/images.go]: to be consistent with the other api, we must use caps.

File: pkg/api/handlers/generic/images.go (hunk 22 above)
Comment (baude) [reply pkg/api/handlers/generic/images.go]: to be consistent with the other api, we must use caps.

File: pkg/api/handlers/generic/images.go (hunk 24 above)
Comment (baude) [reply pkg/api/handlers/generic/images.go]: to be consistent with the other api, we must use caps.

File: pkg/api/handlers/generic/containers_create.go (hunk 11 above)
Comment (baude) [reply pkg/api/handlers/generic/containers_create.go]: to be consistent with the other api, we must use caps.

**Pull request merged**